packages/protocol/contracts/liquidity/GrandaMento.sol

```diff
@@ -0,0 +1,608 @@
+        "Stable token exchange amount not within limits"
+      );
+    }
+
+    // Deposit the assets being sold.
+    IERC20 sellToken = sellCelo ? getGoldToken() : IERC20(stableToken);
+    require(
+      sellToken.transferFrom(msg.sender, address(this), sellAmount),
+      "Transfer in of sell token failed"
+    );
+
+    // Record the proposal.
+    // Add 1 to the running proposal count, and use the updated proposal count as
+    // the proposal ID. Proposal IDs intentionally start at 1.
+    exchangeProposalCount = exchangeProposalCount.add(1);
+    // For stable tokens, the amount is stored in units to deal with demurrage.
+    uint256 storedSellAmount = sellCelo
+      ? sellAmount
+      : IStableToken(stableToken).valueToUnits(sellAmount);
+    exchangeProposals[exchangeProposalCount] = ExchangeProposal({
+      exchanger: msg.sender,
+      stableToken: stableToken,
+      state: ExchangeProposalState.Proposed,
+      sellCelo: sellCelo,
+      sellAmount: storedSellAmount,
+      buyAmount: buyAmount,
+      celoStableTokenExchangeRate: celoStableTokenExchangeRate,
+      approvalTimestamp: 0 // initial value when not approved yet
+    });
+    // Push it into the array of active proposals.
+    activeProposalIdsSuperset.push(exchangeProposalCount);
+    // Even if stable tokens are being sold, the sellAmount emitted is the "value."
+    emit ExchangeProposalCreated(
+      exchangeProposalCount,
+      msg.sender,
+      stableTokenRegistryId,
+      sellAmount,
+      buyAmount,
+      sellCelo
+    );
+    return exchangeProposalCount;
+  }
+
+  /**
+   * @notice Approves an existing exchange proposal.
+   * @dev Sender must be the approver. Exchange proposal must be in the Proposed state.
+   * @param proposalId The identifier of the proposal to approve.
+   */
+  function approveExchangeProposal(uint256 proposalId) external nonReentrant onlyApprover {
+    ExchangeProposal storage proposal = exchangeProposals[proposalId];
+    // Ensure the proposal is in the Proposed state.
+    require(proposal.state == ExchangeProposalState.Proposed, "Proposal must be in Proposed state");
+    // Ensure the change in the current price of CELO quoted in the stable token
+    // relative to the value when the proposal was created is within the allowed limit.
+    FixidityLib.Fraction memory currentRate = getOracleExchangeRate(proposal.stableToken);
+    FixidityLib.Fraction memory proposalRate = FixidityLib.wrap(
+      proposal.celoStableTokenExchangeRate
+    );
+    (FixidityLib.Fraction memory lesserRate, FixidityLib.Fraction memory greaterRate) = currentRate
+      .lt(proposalRate)
+      ? (currentRate, proposalRate)
+      : (proposalRate, currentRate);
+    FixidityLib.Fraction memory rateChange = greaterRate.subtract(lesserRate).divide(proposalRate);
+    require(
+      rateChange.lte(maxApprovalExchangeRateChange),
+      "CELO exchange rate is too different from the proposed price"
+    );
```

Comment:
I guess we don't have to guard against this in both directions? ie if the proposer and approver move ahead with an exchange where the proposal rate is much higher than the current rate, thats probably fine?

Reply:
Might be more complex to implement bc then it'd have to consider if it's a stableToken -> CELO or CELO -> stableToken exchange to divide the exchange rate by 1 or not. Thinking I'll keep it the way it is for now? Lmk if you have an issue with it